In [3]:
import pandas as pd
import numpy as np
import os

In [4]:
# import orders df

df_ords = pd.read_csv(r'C:\Users\18179\Documents\CareerFoundry\InstaCart Basket Analysis\02 Data\Original Data\orders.csv', index_col=False)

In [5]:
#import products df

df_prods = pd.read_csv(r"C:\Users\18179\Documents\CareerFoundry\InstaCart Basket Analysis\02 Data\Original Data\products.csv", index_col = False)

In [7]:
#create test df
df_test = pd.DataFrame()

In [8]:
#create mixed column
df_test['mix'] = ['a', 'b', '1', True]

In [9]:
#check for mixed data
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
#change data type to string
df_test['mix'] = df_test['mix'].astype('str')

In [12]:
df_test

mix
0     a
1     b
2     1
3  True

In [13]:
#Find Missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
#Subset of missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
df_prods.shape

(49693, 5)

In [20]:
#Create clean products df

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [21]:
#Check for row Reduction
df_prods_clean.shape

(49677, 5)

In [22]:
#Check for Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
#Create new DF w/o Duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [25]:
df_prods_clean_no_dups.shape

(49672, 5)

In [27]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

2a.
The glaring data point in the products table is the maximum value for price being $99,999 which is probably a mistake since we are looking at grocery items.

In [28]:
#3a Check for Mixed data in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

4a. There are no mixed columns in the Orders data frame.

In [29]:
#5a. Check for missing values in Orders
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The only missing values occur in the days since prior order column. This is logical since there are customers that place orders on back-to-back days. This probably creates a zero value or missing data in the days_since column. This is a good thing for Instacart because it means they have shop happy customers!

In [7]:
#6a. Address missing values

df_miss = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [8]:
df_miss.head(10)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
11   2168274        2    prior             1          2                 11   
26   1374495        3    prior             1          1                 14   
39   3343014        4    prior             1          6                 11   
45   2717275        5    prior             1          3                 12   
50   2086598        6    prior             1          5                 18   
54   2565571        7    prior             1          3                  9   
75    600894        8    prior             1          6                  0   
79    280530        9    prior             1          1                 17   
83   1224907       10    prior             1          2                 14   

    days_since_prior_order  
0                      NaN  
11                     NaN  
26                     NaN  
39                     NaN  
45                     NaN  
50                     NaN  
54                     NaN  
75                     NaN  
79                     NaN  
83                     NaN

In [34]:
df_miss.tail(10)

order_id  user_id eval_set  order_number  order_dow  \
3420839   2508134   206200    prior             1          5   
3420863     28210   206201    prior             1          6   
3420896    799173   206202    prior             1          2   
3420919   1991850   206203    prior             1          1   
3420925   1438269   206204    prior             1          1   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
3420839                 19                     NaN  
3420863                 14                     NaN  
3420896                 12                     NaN  
3420919                 13                     NaN  
3420925                 11                     NaN  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN

The missing information is valuable in itself. A new column could be made to specify that the zero in the days_since column is not missing data, but this will add uneccesary data to the data frome.

In [10]:
#7a. Check for duplacite values

df_dup_ords = df_ords[df_ords.duplicated()]

In [36]:
df_dup_ords

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

#8a. There are no duplicate rows in the orders data frame. 

In [37]:
#9a. Export 
df_ords.to_csv(r'C:\Users\18179\Documents\CareerFoundry\InstaCart Basket Analysis\02 Data\Orders_Consitency_check.csv')

In [38]:
df_prods_clean_no_dups.to_csv(r'C:\Users\18179\Documents\CareerFoundry\InstaCart Basket Analysis\02 Data\Prods_Consistency_check.csv')

In [9]:
df_miss.shape

(206209, 7)

In [11]:
df_dup_ords.shape

(0, 7)